### DISCUSSION TITLE
Topical Variance

### DISCUSSION THREAD

**Discussion by eschaffn** [MaartenGr/BERTopic#2061] [Q&A]
Hey, I'm wondering if there's a good way to measure a topic's variance.

Variance being how different documents within the topic are relative to each other, and the corpus as a whole. Doing pairwise similarity comparisons is expensive so I'd like to avoid that. I know OCTIS has a diversity measure, but this is calculated over the entire topic model, rather than per topic.

Any ideas?

**Comment by MaartenGr** [comment]
> Doing pairwise similarity comparisons is expensive so I'd like to avoid that. 

I don't think this is actually expensive to do with at most a couple of hundred topic embeddings. Running something like the cosine similarity on the topic embeddings with one another is quite fast and I believe does not take that much memory.

**Reply by eschaffn** [reply]
Maybe to clarify what I'm trying to figure out is: 
If two topics are close in distance to one another (cosine of topic embedding), do they also have significant overlap in the document space? Not sure if this makes sense, but I can try to explain further if not. 

Would these two things always be correlated? If we have two pairs of topics, Topics A and B and Topics C and D, where Cos(A,B) = .8 and Cos(C,D) = .8, could A and B be highly diverse and have significant overlap, while C and D are both coherent and tightly clustered?

**Reply by MaartenGr** [reply]
> If two topics are close in distance to one another (cosine of topic embedding), do they also have significant overlap in the document space? Not sure if this makes sense, but I can try to explain further if not.

The topic embedding by itself doesn't tell you directly something about the structure of the cluster itself, only the centroid.

> If we have two pairs of topics, Topics A and B and Topics C and D, where Cos(A,B) = .8 and Cos(C,D) = .8, could A and B be highly diverse and have significant overlap, while C and D are both coherent and tightly clustered?

Do you mean by "diversity" the distribution of points within a cluster? If so, then the cosine similarity scores will only tell you something about the about a single point (the centroid) rather than the distribution of documents within a cluster.

**Reply by eschaffn** [reply]
Yes this is why I'm looking for a different metic, not the Cosine between centroids. Something that captures the amount of likely overlap between two nearby clusters. 

I want to then merge these clusters based on not only distances of centroids, but also this diversity or variance factor

**Reply by MaartenGr** [reply]
I think options you could take is looking at distribution-based distance/divergence metrics such as Kullback–Leibler divergence, Mahalanobis distance, etc. Having said that, they often assume some sort of probability distribution (which they aren't unless you take the cosine similarity of each point to the centroid).

**Reply by eschaffn** [reply]
Thanks, I'll look into these! 

The distribution of the similarity of each point to the centroid actually sounds very close to what I'm looking for, and avoids pairwise computations.

**Reply by eschaffn** [reply]
If the topic embedding is just the mean of document embeddings of a cluster, this is essentially just the statistical variance then, correct?

**Reply by MaartenGr** [reply]
Yep, that's how I would indeed approach it as statistical variance. There are quite a bit of distance metrics to try out such as the methods shown above, so I would advise just trying some of them out and see whether they make sense for your use case.

**Comment by eschaffn** [comment]
Decided to go with a slightly modified version of: 
https://stackoverflow.com/questions/59919627/how-to-calculate-the-silhouette-score-for-each-cluster-separately-in-python
